### PR TITLE
[`pygrep_hooks`] Fix `blanket-noqa` panic when last line has noqa with no newline (`PGH004`)

### LINKED ISSUE #11102: Rule PGH004 cause panic
ruff 0.4.1 (latest changes from main branch)
```
ruff  *.py --select PGH004 --no-cache --fix --unsafe-fixes --preview --output-format concise --isolated
```

file content:
```
#noqa
```

error
```
All checks passed!

error: Panicked while linting /opt/tmp_folder/10704147344890583060.py: This indicates a bug in Ruff. If you could open an issue at:

    https://github.com/astral-sh/ruff/issues/new?title=%5BLinter%20panic%5D

...with the relevant file contents, the `pyproject.toml` settings, and the following stack trace, we'd be very appreciative!

panicked at /home/runner/work/Automated-Fuzzer/Automated-Fuzzer/ruff/crates/ruff_source_file/src/locator.rs:463:23:
byte index 6 is out of bounds of `#noqa`
Backtrace:    0: ruff::panic::catch_unwind::{{closure}}
             at ./ruff/crates/ruff/src/panic.rs:31:25
   1: <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call
             at /rustc/25ef9e3d85d934b27d9dada2f9dd52b1dc63bb04/library/alloc/src/boxed.rs:2029:9
   2: std::panicking::rust_panic_with_hook
             at /rustc/25ef9e3d85d934b27d9dada2f9dd52b1dc63bb04/library/std/src/panicking.rs:785:13
   3: std::panicking::begin_panic_handler::{{closure}}
             at /rustc/25ef9e3d85d934b27d9dada2f9dd52b1dc63bb04/library/std/src/panicking.rs:659:13
   4: std::sys_common::backtrace::__rust_end_short_backtrace
             at /rustc/25ef9e3d85d934b27d9dada2f9dd52b1dc63bb04/library/std/src/sys_common/backtrace.rs:171:18
   5: rust_begin_unwind
             at /rustc/25ef9e3d85d934b27d9dada2f9dd52b1dc63bb04/library/std/src/panicking.rs:647:5
   6: core::panicking::panic_fmt
             at /rustc/25ef9e3d85d934b27d9dada2f9dd52b1dc63bb04/library/core/src/panicking.rs:72:14
   7: core::str::slice_error_fail_rt
   8: core::str::slice_error_fail
             at /rustc/25ef9e3d85d934b27d9dada2f9dd52b1dc63bb04/library/core/src/str/mod.rs:91:9
   9: ruff_linter::rules::pygrep_hooks::rules::blanket_noqa::blanket_noqa
  10: ruff_linter::checkers::noqa::check_noqa
             at ./ruff/crates/ruff_linter/src/checkers/noqa.rs:208:9
  11: ruff_linter::linter::check_path
             at ./ruff/crates/ruff_linter/src/linter.rs:290:23
  12: ruff_linter::linter::lint_fix
             at ./ruff/crates/ruff_linter/src/linter.rs:550:22
  13: ruff::diagnostics::lint_path
             at ./ruff/crates/ruff/src/diagnostics.rs:280:14
  14: ruff::commands::check::lint_path::{{closure}}
             at ./ruff/crates/ruff/src/commands/check.rs:194:9
  15: std::panicking::try::do_call
             at /rustc/25ef9e3d85d934b27d9dada2f9dd52b1dc63bb04/library/std/src/panicking.rs:554:40
  16: std::panicking::try
             at /rustc/25ef9e3d85d934b27d9dada2f9dd52b1dc63bb04/library/std/src/panicking.rs:518:19
  17: std::panic::catch_unwind
             at /rustc/25ef9e3d85d934b27d9dada2f9dd52b1dc63bb04/library/std/src/panic.rs:142:14
  18: ruff::panic::catch_unwind
             at ./ruff/crates/ruff/src/panic.rs:40:18
  19: ruff::commands::check::lint_path
             at ./ruff/crates/ruff/src/commands/check.rs:193:18
  20: ruff::commands::check::check::{{closure}}
             at ./ruff/crates/ruff/src/commands/check.rs:94:17
  21: <rayon::iter::filter_map::FilterMapFolder<C,P> as rayon::iter::plumbing::Folder<T>>::consume
             at /home/runner/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rayon-1.10.0/src/iter/filter_map.rs:123:36
  22: rayon::iter::plumbing::Folder::consume_iter
             at /home/runner/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rayon-1.10.0/src/iter/plumbing/mod.rs:178:20
  23: rayon::iter::plumbing::Producer::fold_with
             at /home/runner/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rayon-1.10.0/src/iter/plumbing/mod.rs:109:9
  24: rayon::iter::plumbing::bridge_producer_consumer::helper
             at /home/runner/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rayon-1.10.0/src/iter/plumbing/mod.rs:437:13
  25: rayon::iter::plumbing::bridge_producer_consumer
             at /home/runner/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rayon-1.10.0/src/iter/plumbing/mod.rs:396:12
  26: <rayon::iter::plumbing::bridge::Callback<C> as rayon::iter::plumbing::ProducerCallback<I>>::callback
             at /home/runner/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rayon-1.10.0/src/iter/plumbing/mod.rs:372:13
  27: <rayon::slice::Iter<T> as rayon::iter::IndexedParallelIterator>::with_producer
             at /home/runner/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rayon-1.10.0/src/slice/mod.rs:826:9
  28: rayon::iter::plumbing::bridge
             at /home/runner/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rayon-1.10.0/src/iter/plumbing/mod.rs:356:12
  29: <rayon::slice::Iter<T> as rayon::iter::ParallelIterator>::drive_unindexed
             at /home/runner/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rayon-1.10.0/src/slice/mod.rs:802:9
  30: <rayon::iter::filter_map::FilterMap<I,P> as rayon::iter::ParallelIterator>::drive_unindexed
             at /home/runner/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rayon-1.10.0/src/iter/filter_map.rs:46:9
  31: <rayon::iter::fold::Fold<I,ID,F> as rayon::iter::ParallelIterator>::drive_unindexed
             at /home/runner/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rayon-1.10.0/src/iter/fold.rs:59:9
  32: rayon::iter::reduce::reduce
             at /home/runner/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rayon-1.10.0/src/iter/reduce.rs:15:5
  33: rayon::iter::ParallelIterator::reduce
             at /home/runner/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rayon-1.10.0/src/iter/mod.rs:998:9
  34: ruff::commands::check::check
             at ./ruff/crates/ruff/src/commands/check.rs:165:10
  35: ruff::check
             at ./ruff/crates/ruff/src/lib.rs:422:13
  36: ruff::run
             at ./ruff/crates/ruff/src/lib.rs:199:33
  37: ruff::main
             at ./ruff/crates/ruff/src/main.rs:65:11
  38: core::ops::function::FnOnce::call_once
             at /rustc/25ef9e3d85d934b27d9dada2f9dd52b1dc63bb04/library/core/src/ops/function.rs:250:5
  39: std::sys_common::backtrace::__rust_begin_short_backtrace
             at /rustc/25ef9e3d85d934b27d9dada2f9dd52b1dc63bb04/library/std/src/sys_common/backtrace.rs:155:18
  40: std::rt::lang_start::{{closure}}
             at /rustc/25ef9e3d85d934b27d9dada2f9dd52b1dc63bb04/library/std/src/rt.rs:166:18
  41: core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once
             at /rustc/25ef9e3d85d934b27d9dada2f9dd52b1dc63bb04/library/core/src/ops/function.rs:284:13
  42: std::panicking::try::do_call
             at /rustc/25ef9e3d85d934b27d9dada2f9dd52b1dc63bb04/library/std/src/panicking.rs:554:40
  43: std::panicking::try
             at /rustc/25ef9e3d85d934b27d9dada2f9dd52b1dc63bb04/library/std/src/panicking.rs:518:19
  44: std::panic::catch_unwind
             at /rustc/25ef9e3d85d934b27d9dada2f9dd52b1dc63bb04/library/std/src/panic.rs:142:14
  45: std::rt::lang_start_internal::{{closure}}
             at /rustc/25ef9e3d85d934b27d9dada2f9dd52b1dc63bb04/library/std/src/rt.rs:148:48
  46: std::panicking::try::do_call
             at /rustc/25ef9e3d85d934b27d9dada2f9dd52b1dc63bb04/library/std/src/panicking.rs:554:40
  47: std::panicking::try
             at /rustc/25ef9e3d85d934b27d9dada2f9dd52b1dc63bb04/library/std/src/panicking.rs:518:19
  48: std::panic::catch_unwind
             at /rustc/25ef9e3d85d934b27d9dada2f9dd52b1dc63bb04/library/std/src/panic.rs:142:14
  49: std::rt::lang_start_internal
             at /rustc/25ef9e3d85d934b27d9dada2f9dd52b1dc63bb04/library/std/src/rt.rs:148:20
  50: main
  51: <unknown>
  52: __libc_start_main
  53: _start

```
[python_compressed.zip](https://github.com/astral-sh/ruff/files/15074839/python_compressed.zip)

Probably regression from #11053 CC @augustelalande

### PR DESCRIPTION
**Pull request by augustelalande**
## Summary

Resolves #11102

The error stems from these lines
https://github.com/astral-sh/ruff/blob/f5c7a62aa65decb9e286bd65ba17f1a3bd1f91e6/crates/ruff_linter/src/noqa.rs#L697-L702
I don't really understand the purpose of incrementing the last index, but it makes the resulting range invalid for indexing into `contents`.

For now I just detect if the index is too high in `blanket_noqa` and adjust it if necessary.

## Test Plan

Created fixture from issue example.

### CODE REVIEW

File: crates/ruff_linter/src/rules/pygrep_hooks/rules/blanket_noqa.rs [hunk 1]
```diff
@@ -88,8 +88,13 @@ pub(crate) fn blanket_noqa(
 ) {
     for directive_line in noqa_directives.lines() {
         if let Directive::All(all) = &directive_line.directive {
-            let line = locator.slice(directive_line.range);
-            let offset = directive_line.range.start();
+            let line = if directive_line.end() > locator.contents().text_len() {
```
Comment (charliermarsh) [inline crates/ruff_linter/src/rules/pygrep_hooks/rules/blanket_noqa.rs:91]: This seems like a bug, if the `.end()` is greater than the contents... Have you looked into how that happens?

Comment (charliermarsh) [reply crates/ruff_linter/src/rules/pygrep_hooks/rules/blanket_noqa.rs:91]: I see the PR summary, but what happens if we remove it?

Comment (augustelalande) [reply crates/ruff_linter/src/rules/pygrep_hooks/rules/blanket_noqa.rs:91]: Ya it generates a fixture failure with W292_1.py

Comment (augustelalande) [reply crates/ruff_linter/src/rules/pygrep_hooks/rules/blanket_noqa.rs:91]: It seems the code mentioned in the PR summary was added by @MichaReiser so maybe he can shed some light.

**Comment by MichaReiser** [comment]
Uff, I wrote this code like a year ago. But what I did is uncomment that specific code in the `noqa` file and run all tests and the following test now fails

https://github.com/astral-sh/ruff/blob/5849a752235f5384bfef733e64dbfc6a01e30be9/crates/ruff_linter/resources/test/fixtures/pycodestyle/W292_1.py#L1-L2


The problem is that we have fixes reported at the very end of the file, and there must be a way to suppress them. 

The way the noqa lines are found is by searching by a given offset:

https://github.com/astral-sh/ruff/blob/de46a36bbcdc1df4a56fc486ae5c7bcf2c4bbba0/crates/ruff_linter/src/noqa.rs#L722-L734

If we have two lines, than the `previous_line.end() == next_line.start()`. That's why the above methods searches the line using `directive.end() < offset` to avoid that a noqa comment suppresses a violation on the next lines.

I must admit, my "solution" is a hack and it would be nice if we could support end-of-file noqa comments in a better way.

**Comment by augustelalande** [comment]
@charliermarsh I implemented a fix which addresses the issue at its source. Let me know if you like it better.

File: crates/ruff_linter/src/noqa.rs [hunk 2]
```diff
@@ -694,14 +695,17 @@ impl<'a> NoqaDirectives<'a> {
...
+        let last_directive_includes_eof = if let Some(last) = directives.last() {
+            last.range.end() == locator.contents().text_len()
+        } else {
+            false
+        };
 
-        Self { inner: directives }
+        Self {
+            inner: directives,
+            last_directive_includes_eof,
```
Comment (MichaReiser) [inline crates/ruff_linter/src/noqa.rs:707]: Nit: I think I would rename this to `includes_end` to make it clear that the comparison should include the end range.

File: crates/ruff_linter/src/noqa.rs [hunk 3]
```diff
@@ -720,17 +724,26 @@ impl<'a> NoqaDirectives<'a> {
...
+                if self.last_directive_includes_eof && index == self.inner.len() - 1 {
+                    Some(index)
                 } else {
-                    std::cmp::Ordering::Greater
+                    None
                 }
-            })
-            .ok()
+            }
+        }
```
Comment (MichaReiser) [inline crates/ruff_linter/src/noqa.rs:746]: The search then becomes

```
self inner.binary_search_by(|directive| {
	if directive.range.end() < offset {
		std::cmp::Ordering::Less
	} else if directive.range.start() >  offset {
		std::cmp::Ordering::Greater
	} 
	// At this point, end >= offset, start <= offset
	else if !directive.includes_end && directive.range.end() == offset  {
		std::cmp::Ordering::Less // I think or should it be greater?
	} else {
		std::cmp::Ordering::Equal
	}

**Comment by charliermarsh** [comment]
I think it's spurious, I'll re-run it.

**Pull request merged**